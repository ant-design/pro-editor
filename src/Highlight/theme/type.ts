export const THEME_DARK = 'dark';
export const THEME_LIGHT = 'light';

// 主题类型
const tuple = <T extends string[]>(...args: T) => args;
const ThemeTypes = tuple(THEME_DARK, THEME_LIGHT);
export type ThemeType = (typeof ThemeTypes)[number];
