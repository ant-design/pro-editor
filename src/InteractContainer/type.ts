import type { ReactNode } from 'react';

export type InteractionType = 'click' | 'hover';

// export enum InteractStatusEnum {
//   Hover = 'hover',
//   Unhover = 'unhover',
//   Selected = 'selected',
//   UnSelected = 'unSelected',
// }

export type InteractStatusType =
  | 'hover'
  | 'unhover'
  | 'unSelected'
  | 'selected';

// export type InteractStatusType = typeof InteractStatusEnum[keyof typeof InteractStatusEnum];
/**
 * 激活的节点
 */
export interface InteractStatusNode {
  index: number;
  rule: CanvasInteractRule;
}

/**
 * 交互规则
 */
export interface CanvasInteractRule {
  /**
   * 规则id
   */
  id: string;
  /**
   * 用于选择对象
   * 支持使用 类、id 选择器
   */
  selectors: string[];
  /**
   * 支持的交互行为
   */
  actions: InteractionType[];
}

export type ContainerNode = Document | Element;
/**
 * 交互模型
 */
export interface CanvasInteractModel extends CanvasInteractRule {
  elements: Element[];
  /**
   * 容器是否匹配
   */
  containerMatched: boolean;
}

export interface ValidElement {
  /**
   * 元素
   */
  element: HTMLElement;
  /**
   * 选择器
   */
  selector: string;
  /**
   * 序号
   */
  index: number;
}

export interface InteractStatus {
  /**
   * 交互状态的动作
   */
  action?: InteractionType;
  /**
   * 交互状态的动作
   */
  status?: InteractStatusType;
  /**
   * 当前的交互节点
   */
  currentNode?: InteractStatusNode;
  /**
   * 实际激活的节点
   */
  activeNode?: InteractStatusNode;
}

export type GetContainer = (() => HTMLElement) | HTMLElement | string;

export type OnInteractionStatusChange = (status: InteractStatus) => void;

export interface ContextCanvasProps {
  /**
   * 交互规则
   */
  rules: CanvasInteractRule[];
  /**
   * 受控模式的状态
   */
  status?: InteractStatus;
  /**
   * 状态变更回调
   */
  onStatusChange?: OnInteractionStatusChange;
  /**
   * 禁用交互
   */
  disabled?: boolean;
  /**
   * 响应画布式交互的父节点
   */
  getContainer?: GetContainer;
  /**
   * 前缀
   */
  prefixCls?: string;
  /**
   * 子组件
   */
  children?: ReactNode;
}
