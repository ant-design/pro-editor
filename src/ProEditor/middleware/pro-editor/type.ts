import { StateCreator } from 'zustand/vanilla';
import { ConfigPublicAction } from '../../store/slices/config';
import { TakeTwo, Write } from '../types/utils';

export interface ProEditorSetStateAction {
  type: unknown;

  /**
   * 如果记录，那么历史记录的名字是什么
   */
  name?: string;
  /**
   * 是否将操作记录到历史记录中
   */
  recordHistory?: boolean;
}

/**
 * 用于给注入方法添加额外的 proEditor 实例对象
 */
export interface ProEditorMiddlewareInjectMethod<Sa extends unknown[], Sr, T, A> {
  setState<A1 extends string | ProEditorSetStateAction>(...a: [...a: TakeTwo<Sa>, action?: A1]): Sr;

  proEditor: {
    undo?: () => void;
    redo?: () => void;
    // clearStorage: () => void;
    getOptions: () => Partial<ProEditorOptions<T, A>>;
  };
}
/**
 * 提供给用户的配置项
 */
export interface ProEditorOptions<S, EditorSaveState = S> {
  /** Name of the storage (must be unique) */
  name: string;
  /**
   * Filter the persisted value.
   *
   * @params state The state's value
   */
  partialize?: (state: S) => EditorSaveState;
}

// 为  mutator 注入 'pro-editor' 类型，以支持第三个配置参数

type WithProEditor<S, A> = S extends {
  getState: () => infer T;
  setState: (...a: infer Sa) => infer Sr;
}
  ? Write<S, ProEditorMiddlewareInjectMethod<Sa, Sr, T, A>>
  : never;

declare module 'zustand/vanilla' {
  interface StoreMutators<S, A> {
    ['pro-editor']: WithProEditor<S, A>;
  }
}

// 内部方法，用于保证中间件内部的类型定义安全

export type ProEditorImpl = <T>(
  storeInitializer: StateCreator<T, [['pro-editor', unknown]], []>,
  options: ProEditorOptions<T, T>,
) => StateCreator<T, [['pro-editor', unknown]], []>;

export interface InjectInternalProEditor {
  proEditor: {
    __INTERNAL_SET_CONFIG__NOT_USE_IT: ConfigPublicAction['setConfig'];
  };
}
