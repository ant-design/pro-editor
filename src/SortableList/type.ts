import type { UniqueIdentifier } from '@dnd-kit/core';
import type { ReactNode } from 'react';

export interface KeyManager {
  /**
   * 列表 ID 项
   */
  keys: UniqueIdentifier[];
  /**
   * 当前自增 ID
   */
  id: number;
}

export interface CreatorButtonProps {
  /**
   * 列表有值时是否展示添加按钮
   */
  showInList?: boolean;
  /**
   * 空数据时是否展示添加按钮
   */
  showInEmpty?: boolean;
  /**
   * 生成初始值逻辑
   */
  record?: (index: number) => Record<string, any>;
  /**
   * 新增一行按钮文案
   */
  creatorButtonText?: string;
}

export type RenderItemProps<T = any> = (item: T, index: number) => ReactNode;

export type RenderActionProps<T = any> =
  | ((item: T, index: number) => ReactNode[])
  | React.ReactNode[];

export type { UniqueIdentifier };
