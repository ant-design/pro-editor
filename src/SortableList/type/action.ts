// 新增节点
export interface AddItemAction {
  /**
   * 动作类型
   */
  type: 'addItem';
  /**
   * 新增的节点
   */
  item: any;
  /**
   * 新增节点的位置
   * @default undefined
   */
  index?: number;
}

// 移动节点
export interface MoveItemAction {
  /**
   * 动作类型
   */
  type: 'moveItem';
  /**
   * 当前节点索引
   */
  activeIndex: number;
  /**
   * 目标节点索引
   */
  overIndex: number;
}

// 移除节点
export interface RemoveItemAction {
  /**
   * 动作类型
   */
  type: 'removeItem';
  /**
   * 要移除的节点的位置
   */
  index: number;
}

// 修改节点
export interface UpdateItemAction {
  /**
   * 动作类型
   */
  type: 'updateItem';
  /**
   * 要修改的节点的位置
   */
  index: number;
  /**
   * 修改后的节点内容
   */
  item: any;
}

/**
 * 内部 Item 数据更新方法
 */
export type SortableListDispatchPayload =
  | MoveItemAction
  | AddItemAction
  | RemoveItemAction
  | UpdateItemAction;
