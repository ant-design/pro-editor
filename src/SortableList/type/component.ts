import type { DraggableSyntheticListeners, UniqueIdentifier } from '@dnd-kit/core';
import type { Transform } from '@dnd-kit/utilities';
import type { CSSProperties, FC, ReactElement, Ref } from 'react';
import { StoreUpdaterProps } from '../type/store';

export interface BaseItemProps<T = any>
  extends Pick<StoreUpdaterProps<T>, 'renderItem' | 'renderContent' | 'actions' | 'hideRemove'> {
  dragOverlay?: boolean;
  color?: string;
  disabled?: boolean;
  dragging?: boolean;
  handle?: boolean;
  height?: number;
  item: T;
  index?: number;
  fadeIn?: boolean;
  transform?: Transform | null;
  listeners?: DraggableSyntheticListeners;
  sorting?: boolean;
  className?: string;
  style?: React.CSSProperties;
  transition?: string | null;
  id: UniqueIdentifier;
  onRemove?: () => void;
}

export type RenderActionProps<T = any> =
  | ((item: T, index: number) => React.ReactNode[])
  | React.ReactNode[];

export interface SortableItemProps<T = any>
  extends Pick<
    StoreUpdaterProps<T>,
    'renderItem' | 'renderContent' | 'actions' | 'getItemStyles' | 'hideRemove' | 'handle'
  > {
  disabled?: boolean;
  id: UniqueIdentifier;
  index: number;
  item: T;
  useDragOverlay?: boolean;
  onRemove?: (index: number) => void;
}

export type RenderContent<T = any> = (item: T, index: number) => React.ReactNode;

export type RenderEmpty = () => ReturnType<FC>;

export type RenderHeader = () => ReturnType<FC>;

export interface CreatorButtonProps {
  /**
   * 添加创建按钮的位置
   */
  position?: 'bottom' | 'top';
  /**
   * 生成初始值逻辑
   */
  record: (index: number) => any;
  /**
   * 新增一行按钮文案
   */
  creatorButtonText?: string;
  /**
   * 按钮的样式设置，可以设置按钮是否显示，比如 style: {display: 'none'}
   */
  style?: CSSProperties;
}

export type RenderItem<T = any> = (
  item: T,
  options: {
    /**
     * 是否是被拖出的列表项
     */
    dragOverlay: boolean;
    /**
     * 是否在拖拽中
     */
    dragging: boolean;
    /**
     * 是否在排序中
     */
    sorting: boolean;
    /**
     * 当前列表项索引
     */
    index: number | undefined;
    /**
     * fade 动画
     */
    fadeIn: boolean;
    /**
     * 拖拽项监听器
     */
    listeners: DraggableSyntheticListeners;
    /**
     * ref
     */
    ref: Ref<HTMLElement>;
    /**
     * 当面列表项传入样式
     */
    style: CSSProperties | undefined;
    /**
     * 当面列表项 transform 动画
     */
    transform: any;
    /**
     * 当面列表项 transition 动画
     */
    transition: any;
  },
) => ReactElement;

export type { UniqueIdentifier };

interface GetItemStylesArgs {
  /**
   * 当前列表项索引
   */
  index: number;
  /**
   * 是否在拖拽中
   */
  isDragging: boolean;
  /**
   * 当前列表项 ID
   */
  id: UniqueIdentifier;
  /**
   * 是否在排序中
   */
  isSorting: boolean;
  /**
   * 拖拽覆盖的列表项索引
   */
  overIndex: number;
  /**
   * 是否是拖拽中的列表项
   */
  isDragOverlay: boolean;
}

export type GetItemStyles = (status: GetItemStylesArgs) => CSSProperties;
