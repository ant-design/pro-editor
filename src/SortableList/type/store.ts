import { ForwardedRef } from 'react';
import type {
  CreatorButtonProps,
  GetItemStyles,
  RenderActionProps,
  RenderContent,
  RenderEmpty,
  RenderItem,
  SortableListDispatchPayload,
  UniqueIdentifier,
} from '../type';

export type OnChange<T = any> = (items: T[], event: SortableListDispatchPayload) => void;

export interface SortableListState {
  /*
   * 值变化
   */
  onChange?: OnChange;
  /**
   * 删除回调
   */
  onRemove?: (index: number) => void;
  /**
   * 当前选中项 ID
   */
  activeId?: UniqueIdentifier;
  /**
   * 值
   */
  value?: any[];
  /**
   * 和 Value 值对应的 id 管理器
   */
  keyManager?: UniqueIdentifier[];
  /**
   * 渲染额外区域
   */
  actions?: RenderActionProps;
  /**
   * 渲染可排序项
   */
  renderItem?: RenderItem;
  /**
   * 渲染空状态
   */
  renderEmpty?: RenderEmpty;
  /**
   * 渲染可排序项内容
   */
  renderContent?: RenderContent;
  /**
   * 新建对象相关属性
   */
  creatorButtonProps?: CreatorButtonProps | false;
  /**
   * 渲染可排序列表项样式
   */
  getItemStyles?: GetItemStyles;
  /**
   * 隐藏
   */
  hideRemove?: boolean;
}

/**
 * 供外部使用的 ref 方法
 */
export interface SortableListRef<T = any> {
  addItem: (item?: T, index?: number) => void;
  removeItem: (index: number) => void;
  updateItem: (item: T, index: number) => void;
}

export interface StoreUpdaterProps<T = any> {
  /**
   * 值
   */
  value?: T[];
  /**
   * 初始值
   */
  initialValues?: T[];
  /**
   * 值变化
   */
  onChange?: OnChange<T>;
  /**
   * 渲染可排序项
   */
  renderItem?: RenderItem<T>;
  /**
   * 渲染内容区域
   */
  renderContent?: RenderContent<T>;
  /**
   * 渲染空状态
   */
  renderEmpty?: RenderEmpty;
  /**
   * 新建对象相关属性
   */
  creatorButtonProps?: CreatorButtonProps | false;
  /**
   * 除列表自带操作之外的其他操作自渲染
   */
  actions?: RenderActionProps<T>;
  /**
   * 对外部暴露方法
   */
  ref?: ForwardedRef<SortableListRef<T>>;
  /**
   * 如果为 true，则会在 devtools 中显示 SortableTree 内部的数据结构
   * @internal
   */
  SHOW_STORE_IN_DEVTOOLS?: boolean;
  /**
   * 是否隐藏删除按钮，默认为 false
   */
  hideRemove?: boolean;
  /**
   * 自定义排序容器样式
   */
  getItemStyles?: GetItemStyles;
}
