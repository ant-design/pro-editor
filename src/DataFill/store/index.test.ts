import { act, renderHook } from '@testing-library/react';
import type { FilledPayload } from '../types';
import { createStore } from './index';

const useStore = createStore();

describe('useStore', () => {
  it('默认快照', () => {
    const { result } = renderHook(() => useStore());

    expect(result.current).toMatchSnapshot('default');
  });

  it('修改配置项后触发填充', () => {
    const { result } = renderHook(() => useStore());
    let fillFunc;

    const onItemClick = (payload: FilledPayload) => {
      fillFunc = payload.filler;
    };

    // 默认脱敏
    act(() => {
      result.current.updateState({ onItemClick });
      result.current.handleFill('idCard');
    });

    expect(fillFunc()).toHaveLength(12);

    // 修改为不脱敏
    act(() => {
      result.current.updateConfig({ idCard: false });
      result.current.handleFill('idCard');
    });

    expect(fillFunc()).toHaveLength(18);
  });

  it('受控模式：触发onChange', () => {
    const { result } = renderHook(() => useStore());

    // 设置 onChange 方法
    const keyMap = [];
    const onItemClick = (payload: FilledPayload) => {
      keyMap.push(payload.key);
    };

    act(() => {
      result.current.updateState({ onItemClick });
    });
    expect(result.current.onItemClick).toEqual(onItemClick);

    // 更新 data
    act(() => {
      result.current.handleFill('name');
    });

    expect(keyMap).toEqual(['name']);

    act(() => {
      result.current.handleFill('time');
    });
    expect(keyMap).toEqual(['name', 'time']);
  });
});
