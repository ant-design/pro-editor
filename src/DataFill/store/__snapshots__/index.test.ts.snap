// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useStore > 默认快照 > default 1`] = `
{
  "config": {
    "bankCard": true,
    "data": "hundreds",
    "dateTime": "dateTime",
    "idCard": true,
    "landline": true,
    "month": "zh",
    "name": "zh",
    "percent": false,
    "phone": true,
    "website": "url",
    "week": "zh",
  },
  "fillTreeData": [
    {
      "children": [
        {
          "icon": <ForwardRef(UserOutlined) />,
          "key": "name",
          "params": [
            {
              "label": "中文",
              "value": "zh",
            },
            {
              "label": "英文",
              "value": "en",
            },
          ],
          "title": "人名",
        },
        {
          "icon": <ForwardRef(GlobalOutlined) />,
          "key": "website",
          "params": [
            {
              "label": "链接",
              "value": "url",
            },
            {
              "label": "域名",
              "value": "domain",
            },
            {
              "label": "IP",
              "value": "ip",
            },
          ],
          "title": "网址",
        },
        {
          "icon": <ForwardRef(MailOutlined) />,
          "key": "mail",
          "title": "邮件",
        },
      ],
      "icon": <FillText />,
      "key": "common",
      "params": [
        {
          "label": "短文本",
          "value": "default",
        },
        {
          "label": "长文本",
          "value": "longText",
        },
      ],
      "title": "文本",
    },
    {
      "children": [
        {
          "checked": {
            "label": "脱敏",
            "value": false,
          },
          "icon": <ForwardRef(PhoneOutlined) />,
          "key": "phone",
          "title": "手机",
        },
        {
          "checked": {
            "label": "脱敏",
            "value": false,
          },
          "icon": <ForwardRef(PhoneOutlined) />,
          "key": "landline",
          "title": "座机",
        },
        {
          "checked": {
            "label": "脱敏",
            "value": false,
          },
          "icon": <ForwardRef(IdcardOutlined) />,
          "key": "idCard",
          "title": "身份证",
        },
        {
          "checked": {
            "label": "脱敏",
            "value": false,
          },
          "icon": <ForwardRef(IdcardOutlined) />,
          "key": "bankCard",
          "title": "信用卡/借记卡",
        },
      ],
      "key": "identifier",
      "title": "编号",
    },
    {
      "children": [
        {
          "checked": {
            "label": "小数",
            "value": false,
          },
          "icon": <ForwardRef(PercentageOutlined) />,
          "key": "percent",
          "title": "百分比",
        },
        {
          "checked": {
            "label": "小数",
            "value": false,
          },
          "icon": <ForwardRef(DollarOutlined) />,
          "key": "money",
          "title": "金额",
        },
      ],
      "icon": <FillNumber />,
      "key": "data",
      "params": [
        {
          "label": "0~9",
          "value": "digits",
        },
        {
          "label": "10~99",
          "value": "tens",
        },
        {
          "label": "100~999",
          "value": "hundreds",
        },
      ],
      "title": "数据",
    },
    {
      "children": [
        {
          "icon": <ForwardRef(CalendarOutlined) />,
          "key": "date",
          "params": [
            {
              "label": "年月日",
              "value": "yyyy年MM月dd日",
            },
            {
              "label": "年/月/日",
              "value": "yyyy/MM/dd",
            },
            {
              "label": "年-月-日",
              "value": "yyyy-MM-dd",
            },
          ],
          "title": "日期",
        },
        {
          "icon": <FillMonth />,
          "key": "month",
          "params": [
            {
              "label": "中文",
              "value": "zh",
            },
            {
              "label": "英文",
              "value": "en",
            },
            {
              "label": "英文简写",
              "value": "abbr",
            },
          ],
          "title": "月份",
        },
        {
          "icon": <FillWeek />,
          "key": "week",
          "params": [
            {
              "label": "中文",
              "value": "zh",
            },
            {
              "label": "英文",
              "value": "en",
            },
            {
              "label": "英文简写",
              "value": "abbr",
            },
          ],
          "title": "星期",
        },
      ],
      "icon": <ForwardRef(ClockCircleOutlined) />,
      "key": "dateTime",
      "params": [
        {
          "label": "日期时间",
          "value": "dateTime",
        },
        {
          "label": "日期",
          "value": "date",
        },
        {
          "label": "时间",
          "value": "time",
        },
      ],
      "title": "时间",
    },
    {
      "children": [
        {
          "icon": <FillCountry />,
          "key": "country",
          "title": "国家和地区",
        },
        {
          "key": "province",
          "title": "省份",
        },
        {
          "icon": <FillCity />,
          "key": "city",
          "title": "城市",
        },
      ],
      "icon": <ForwardRef(EnvironmentOutlined) />,
      "key": "address",
      "title": "地址",
    },
  ],
  "handleFill": [Function],
  "handleShowDemoData": [Function],
  "locale": "zh_CN",
  "updateConfig": [Function],
  "updateState": [Function],
}
`;
